src/creds.rs

```diff
@@ -185,6 +196,53 @@ impl Codec for BasicCredential {
     // }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct ExtensibleCredential {
+    pub identity: Vec<u8>,
+    pub ciphersuite: Ciphersuite,
+    pub public_key: SignaturePublicKey,
+    pub extensions: Vec<Extension>,
+}
+
+impl ExtensibleCredential {
+    pub fn verify(&self, payload: &[u8], signature: &Signature) -> bool {
+        self.ciphersuite
+            .verify(signature, &self.public_key, payload)
+    }
+}
+
+//impl From<&Identity, &Vec<Extension>> for ExtensibleCredential {
+//    fn from(identity: &Identity, extensions: &Vec<Extension>) -> Self {
+//        ExtensibleCredential {
+//            identity: identity.id.clone(),
+//            ciphersuite: identity.ciphersuite,
+//            public_key: identity.keypair.get_public_key().clone(),
+//        }
+//    }
+//}
+
+impl Codec for ExtensibleCredential {
+    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
+        encode_vec(VecSize::VecU16, buffer, &self.identity)?;
+        self.ciphersuite.encode(buffer)?;
+        self.public_key.encode(buffer)?;
+        encode_vec(VecSize::VecU32, buffer, &self.extensions)?;
+        Ok(())
+    }
+    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
```

Comment:
Please don't implement `decode` if it's not absolutely necessary.

Reply:
Fair enough. The reason being that we need encode for signatures, but decode only if we expect to receive it from the network?

Reply:
I want to get rid of almost all decodes. Right, it's only to build the object from bytes. And this will only work as long as it's a simple object that doesn't have any references to other objects.

Reply:
I see. I think a lot of the time, the objects are 1-to-1 implementations of the structs we have in the spec, so it should be possible for most of them. But of course that doesn't always make sense.